sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/variant/variantExpressions.scala

```diff
@@ -53,3 +66,320 @@ case class ParseJson(child: Expression)
+      DataTypeMismatch(
+        errorSubClass = "NON_FOLDABLE_INPUT",
+        messageParameters = Map(
+          "inputName" -> toSQLId("path"),
+          "inputType" -> toSQLType(path.dataType),
+          "inputExpr" -> toSQLExpr(path)))
+    } else if (!VariantGet.checkDataType(targetType)) {
+      DataTypeMismatch(
+        errorSubClass = "CAST_WITHOUT_SUGGESTION",
+        messageParameters =
+          Map("srcType" -> toSQLType(VariantType), "targetType" -> toSQLType(targetType)))
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override lazy val dataType: DataType = targetType.asNullable
+
+  @transient private lazy val parsedPath = {
+    val pathValue = path.eval().toString
+    VariantPathParser.parse(pathValue).getOrElse {
+      throw QueryExecutionErrors.invalidVariantGetPath(pathValue, prettyName)
+    }
+  }
+
+  final override def nodePatternsInternal(): Seq[TreePattern] = Seq(VARIANT_GET)
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(VariantType, StringType)
+
+  override def prettyName: String = if (failOnError) "variant_get" else "try_variant_get"
+
+  override def nullable: Boolean = true
+
+  protected override def nullSafeEval(input: Any, path: Any): Any = {
+    VariantGet.variantGet(
+      input.asInstanceOf[VariantVal],
+      parsedPath,
+      dataType,
+      failOnError,
+      timeZoneId)
+  }
+
+  override def left: Expression = child
+
+  override def right: Expression = path
+
+  override protected def withNewChildrenInternal(
+      newChild: Expression,
+      newPath: Expression): VariantGet = copy(child = newChild, path = newPath)
+
+  override def withTimeZone(timeZoneId: String): VariantGet = copy(timeZoneId = Option(timeZoneId))
+}
+
+case object VariantGet {
+  /**
+   * Returns whether a data type can be cast into/from variant. For scalar types, we allow a subset
+   * of them. For nested types, we reject map types with a non-string key type.
+   */
+  def checkDataType(dataType: DataType): Boolean = dataType match {
+    case _: NumericType | BooleanType | _: StringType | BinaryType | TimestampType | DateType |
+        VariantType =>
+      true
+    case ArrayType(elementType, _) => checkDataType(elementType)
+    case MapType(StringType, valueType, _) => checkDataType(valueType)
+    case StructType(fields) => fields.forall(f => checkDataType(f.dataType))
+    case _ => false
+  }
+
+  /** The actual implementation of the `VariantGet` expression. */
+  def variantGet(
+      input: VariantVal,
+      parsedPath: Array[VariantPathParser.PathSegment],
+      dataType: DataType,
+      failOnError: Boolean,
+      zoneId: Option[String]): Any = {
+    var v = new Variant(input.getValue, input.getMetadata)
+    for (path <- parsedPath) {
+      v = path match {
+        case scala.util.Left(key) if v.getType == Type.OBJECT => v.getFieldByKey(key)
+        case scala.util.Right(index) if v.getType == Type.ARRAY => v.getElementAtIndex(index)
+        case _ => null
```

Comment:
We can remove this case match. This pattern match is exhausted.

Reply:
It is not really exhausted due to `if v.getType == Type.OBJECT/ARRAY`. For example, if the variant type is a `LONG`, we should return null. Or if the variant type is `ARRAY` and the `path` is an object field access, we should also return null.